### PR TITLE
[Drop-In UI] Added support for Info Panel layout customization.

### PR DESCRIPTION
**Pull request by tomaszrybakiewicz**
Closes [#1824](https://github.com/mapbox/navigation-sdks/issues/1824)

### TL;DR
This PR updates how Info Panel is laid out in the NavigationView. It introduces InfoPanelBinder which can be used to inject a custom info panel layout.

### Description

Changes:
- Introduced `ViewBinderCustomization.infoPanelBinder` that can be used to inject custom info panel binder implementation.
- Refactored `InfoPanelBinder` into an abstract, base binder class that must be subclassed to provide a custom info panel layout.
- Introduced `ViewStyleCustomization.infoPanelPeekHeight` that allows change of bottom sheet peek height.
- Updated InfoPanelCoordinator to support new customization values.
- Updated QA Test App "Customized NavigationView" activity to showcase new info panel customizations.

### Screenshots or Gifs

_(Recording of QA Test App captured on Pixel 2)_

https://user-images.githubusercontent.com/2678039/178601130-560c4cd3-48d4-4f7c-a409-b1078aad6ec1.mp4

### CODE REVIEW

**Comment by Zayankovsky** [comment]
Do I understand correctly that if one uses `InfoPanelBinder`, they will no longer have the default info panel header that comes with Drop-In UI?

File: libnavui-dropin/src/main/res/layout/mapbox_info_panel_layout.xml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:orientation="vertical"
+    android:background="@drawable/mapbox_bg_info_panel"
+    android:elevation="8dp">
+
+    <FrameLayout
+        android:id="@+id/infoPanelHeader"
+        android:layout_width="match_parent"
+        android:layout_height="@dimen/mapbox_infoPanel_peekHeight"
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/res/layout/mapbox_info_panel_layout.xml:15]: I think it should be okay to set it to `wrap_content`. Doing that results in the screenshot attached below.




Maybe I am missing something obvious, but I think this is an issue. Because, in this case we are adding a new layout to `header binder` in the form of `qa_test_app/res/layout_info_panel_header` which has a `constraintLayout` with height of `104dp`. So how come the above `FrameLayout` `height` doesn't expand to `104dp`, but rather gets limited to the height of `textView`?

File: libnavui-dropin/src/main/res/layout/mapbox_navigation_view_layout.xml [hunk 2]
```diff
@@ -122,29 +122,23 @@
...
-                android:orientation="vertical"
                 android:layout_width="match_parent"
                 android:layout_height="wrap_content"
-                android:elevation="8dp"
-                android:background="@drawable/mapbox_bg_info_panel"
                 app:layout_behavior="com.google.android.material.bottomsheet.BottomSheetBehavior"
                 app:behavior_hideable="true"
                 app:behavior_fitToContents="true"
-                app:behavior_peekHeight="@dimen/mapbox_infoPanel_peekHeight">
+                app:behavior_peekHeight="@dimen/mapbox_infoPanel_peekHeight"
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/res/layout/mapbox_navigation_view_layout.xml:132]: We don't need it anymore since we are assigning peek height at runtime. I removed it on your branch locally and it's working fine.

File: libnavui-dropin/src/main/res/layout/mapbox_navigation_view_layout.xml [hunk 3]
```diff
@@ -122,29 +122,23 @@
...
-                    android:layout_height="@dimen/mapbox_infoPanel_peekHeight" />
-
-                <FrameLayout
-                    android:id="@+id/infoPanelContent"
-                    android:layout_width="match_parent"
-                    android:layout_height="wrap_content" />
+                <!--
+                    InfoPanelBinder will inject customizable content container
+                    which will host mapbox_info_panel_layout content.
+                -->
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/res/layout/mapbox_navigation_view_layout.xml:139]: I don't think we need this comment here. It's pretty obvious with Drop-In UI architecture that views will be injected to these frame layouts later.

File: libnavui-dropin/src/main/res/layout-land/mapbox_navigation_view_layout.xml [hunk 4]
```diff
@@ -129,29 +129,23 @@
...
-                android:orientation="vertical"
                 android:layout_width="375dp"
                 android:layout_height="wrap_content"
-                android:elevation="8dp"
-                android:background="@drawable/mapbox_bg_info_panel"
                 app:layout_behavior="com.google.android.material.bottomsheet.BottomSheetBehavior"
                 app:behavior_hideable="true"
                 app:behavior_fitToContents="true"
-                app:behavior_peekHeight="@dimen/mapbox_infoPanel_peekHeight">
+                app:behavior_peekHeight="@dimen/mapbox_infoPanel_peekHeight"
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/res/layout-land/mapbox_navigation_view_layout.xml:139]: We don't need it anymore since we are assigning peek height at runtime. I removed it on your branch locally and it's working fine.

File: libnavui-dropin/src/main/res/layout-land/mapbox_navigation_view_layout.xml [hunk 5]
```diff
@@ -129,29 +129,23 @@
...
-                    android:layout_height="@dimen/mapbox_infoPanel_peekHeight" />
-
-                <FrameLayout
-                    android:id="@+id/infoPanelContent"
-                    android:layout_width="match_parent"
-                    android:layout_height="wrap_content" />
+                <!--
+                    InfoPanelBinder will inject customizable content container
+                    which will host mapbox_info_panel_layout content.
+                -->
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/res/layout-land/mapbox_navigation_view_layout.xml:146]: I don't think we need this comment here. It's pretty obvious with Drop-In UI architecture that views will be injected to these frame layouts later.

**Comment by abhishek1508** [comment]
> Do I understand correctly that if one uses InfoPanelBinder, they will no longer have the default info panel header that comes with Drop-In UI?

With the changes in this PR, customers can follow several different ways:
- Use the default `InfoPanelBinder` which will use the default `InfoPanelHeaderBinder` and `InfoPanelContentBinder`. 
- Use the default `InfoPanelBinder` with custom `InfoPanelHeaderBinder` or `InfoPanelContentBinder` or both. This will not change the look of parent bottom sheet, but you can change the look and feel of header and content details.
- Use custom `InfoPanelBinder` with default `InfoPanelHeaderBinder` and `InfoPanelContentBinder`. This will allow you to change the look of parent info panel.
- Use custom `InfoPanelBinder` with custom `InfoPanelHeaderBinder` or `InfoPanelContentBinder` or both. 

/cc @Zayankovsky

**Comment by abhishek1508** [comment]
I am seeing the following leak

File: qa-test-app/src/main/res/layout/layout_drawer_menu_nav_view_custom.xml [hunk 6]
```diff
@@ -69,13 +69,31 @@
...
+        <TableRow android:padding="5dp">
+            <androidx.appcompat.widget.SwitchCompat
+                android:id="@+id/toggleCustomInfoPanel"
+                android:layout_weight="1"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:text="Custom Info Panel Layout"
+                android:textAllCaps="true" />
+        </TableRow>
+        <TableRow android:padding="5dp">
```
Comment (abhishek1508) [inline qa-test-app/src/main/res/layout/layout_drawer_menu_nav_view_custom.xml:81]: I also think we should remove this toggle switch completely. It's a bit annoying that I see the `InfoPanel` and it's asking me to `pull up for details`; but when I try to it doesn't work because the functionality is disabled.

**Comment by Zayankovsky** [comment]
@tomaszrybakiewicz @abhishek1508 I'm trying to add search UI to the info panel in Free Drive state. At the moment in order to achieve this I will have to split the Search view into two parts (header and content) which is rather inconvenient, especially since the header height is fixed. At the same time I would like to keep default Drop-In content (without having to reimplement it) in the info panel header during other states, so I'm fine with this splitting into header and content parts when not in Free Drive.
So my question is whether it is going to be possible to achieve that with the changes introduced in this PR (display a single view during Free Drive, and customize info panel content while keeping default info panel header during other states) and what should be the proper way to implement that (which binders to use and when).

File: libnavui-dropin/src/main/res/layout/mapbox_info_panel_layout.xml (hunk 1 above)
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/res/layout/mapbox_info_panel_layout.xml:15]: In what OS version did you notice this?
I have tested this in an _emulator with Android 6.0 (API 23)_, a  _device with Android 11 (API 30)_  and a _device with Android 12 (API 32)_. It works as expected.
|emulator with Android 6.0 <br/>(API 23)| Google Pixel 2 Android 11 <br/>(API 30)  | Google Pixel 6 Android 12 <br/>(API 32) |
|:---:|:---:|:---:|
|emulator-android6-api23 | pixel2 | pixel6 |

**Comment by abhishek1508** [comment]
> I'm trying to add search UI to the info panel in Free Drive state. At the moment in order to achieve this I will have to split the Search view into two parts (header and content) which is rather inconvenient, especially since the header height is fixed. At the same time I would like to keep default Drop-In content (without having to reimplement it) in the info panel header during other states, so I'm fine with this splitting into header and content parts when not in Free Drive.
So my question is whether it is going to be possible to achieve that with the changes introduced in this PR (display a single view during Free Drive, and customize info panel content while keeping default info panel header during other states) and what should be the proper way to implement that (which binders to use and when).

In regards to 

> especially since the header height is fixed. 

This has already been captured in the comment https://github.com/mapbox/mapbox-navigation-android/pull/6049#discussion_r921485088

This is how the bottom sheet looks like in 1TAP

| FreeDrive | RoutePreview | ActiveNavigation |
| ------ | ------ | ------ |
|  |  |  |

To achieve bottom sheet:
- with margins on both the ends so that it does not take up the entire width
- with a small top left and top right corner radius'

you will have to create your own `InfoPanelBinder` 
@tomaszrybakiewicz This is where I think this is too much to ask and we should have some simple attributes that should allow customers to be able to achieve this without writing an `InfoPanelBinder`

> At the same time I would like to keep default Drop-In content (without having to reimplement it) in the info panel header during other states, so I'm fine with this splitting into header and content parts when not in Free Drive.

Based on your comment above it seems you want to use default `InfoPanelContentBinder` which is always empty. That means for each of the image above and the views they host, you want to add them to the header. To have this views inside header, you would have to write a few `InfoPanelHeaderBinder` since your header views are changing in every state. 

Create one header each for:
- FreeDrive: FreeDriveHeaderBinder
- RoutePreview: RoutePreviewHeaderBinder
- ActiveNavigation: ActiveNavigationHeaderBinder

and inject them to Drop-In UI.

So, to summarize
- Create `1TAPInfoPanelBinder` to have the bottom sheet look like as it does in 1TAP.
- Create `FreeDriveHeaderBinder`, `RoutePreviewHeaderBinder`, `ActiveNavigationHeaderBinder` that would host different views in each of these different states

Inject all these binders to `navigationView`

File: libnavui-dropin/src/main/res/layout/mapbox_info_panel_layout.xml (hunk 1 above)
Comment (abhishek1508) [reply libnavui-dropin/src/main/res/layout/mapbox_info_panel_layout.xml:15]: @tomaszrybakiewicz You tested it with `FrameLayout` height set to `wrap_content` instead of `@dimen/mapbox_infoPanel_peekHeight`. The image I shared above happens when I set the height of the `FrameLayout` to `wrap_content`

Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/res/layout/mapbox_info_panel_layout.xml]: OK. I figured out what's going on here.

I forgot to pass **parent layout** [when inflating header view in the custom InfoPanelHeaderBinder](https://github.com/mapbox/mapbox-navigation-android/pull/6049/files#diff-bad7504047ae8e24d4e03581ce6693a1eb7d3a4b6831e94abe071f2f89d8c9f5R381).
This resulted in an ambiguous layout where a child with `android:layout_height="match_parent"` is  added to a parent with `android:layout_height="wrap_content"`.

I patched it up. Now it should work as expected.

File: libnavui-dropin/src/main/res/layout/mapbox_info_panel_layout.xml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        app:layout_constraintStart_toStartOf="parent"
+        tools:background="#ccffcc"
+        tools:minHeight="200dp" />
+
+    <View
+        style="@style/DropInStyleDragHandle"
+        android:id="@+id/handle"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent" />
```
Comment (tomaszrybakiewicz) [inline libnavui-dropin/src/main/res/layout/mapbox_info_panel_layout.xml:37]: @abhishek1508 Just FYI. I moved this drag handle view from the header view.
This will allow developers to:
- replace or hide it in their custom info panel layout while using the default header
- change the position of the `infoPanelHeader` layout (eg. when placing an additional view above header view)

**Comment by tomaszrybakiewicz** [comment]
> 👍🏻 to adding changes that allows injecting custom info panel although the plan was to introduce this through a different ticket. It's an interesting approach to allow users to:
> 
> add margins
> change background drawables and colors
> by introducing InfoPanelBinder that would enable them to change the design. However, don't you think it's a bit too much of an ask? If I just simply want to have an info panel that does not have rounded corners, with this approach as a customer I will have to write a new InfoPanelBinder.
> 
> My idea was to expose a few settings that could be applied to the default info panel as stated in the OP without asking users to create a new InfoPanelBinder.
> 
> This doesn't mean that we revert the changes introduced in the PR already. We just add a few attributes to the info panel like margins, background drawables and background colors.

@abhishek1508 I don't think this is asking too much. I think the benefits of the introduced solution here outweigh the "ask". 

With the new capability, developers will not only be able to customize the info panel as described in the attached issue, but will also be able to customize how the header and content views are laid out.

Just to show how little developers need to write, here an example.

```xml
<!-- layout_info_panel.xml -->
<?xml version="1.0" encoding="utf-8"?>
<androidx.constraintlayout.widget.ConstraintLayout
    xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    xmlns:tools="http://schemas.android.com/tools"
    android:layout_width="match_parent"
    android:layout_height="wrap_content"
    android:layout_margin="10dp"
    android:background="@drawable/bg_custom_info_panel"
    android:elevation="0dp">

    <FrameLayout
        android:id="@+id/infoPanelHeader"
        android:layout_width="match_parent"
        android:layout_height="@dimen/mapbox_infoPanel_peekHeight"
        app:layout_constraintTop_toTopOf="parent"
        app:layout_constraintEnd_toEndOf="parent"
        app:layout_constraintStart_toStartOf="parent"
        tools:background="#00ccff" />

    <FrameLayout
        android:id="@+id/infoPanelContent"
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        tools:background="#ccffcc"
        tools:minHeight="200dp"
        app:layout_constraintTop_toBottomOf="@id/infoPanelHeader"
        app:layout_constraintEnd_toEndOf="parent"
        app:layout_constraintStart_toStartOf="parent" />

</androidx.constraintlayout.widget.ConstraintLayout>
```

And the `CustomInfoPanelBinder` that injects and describes a customized `layout_info_panel` layout:

```kotlin
private fun customizeInfoPanel(navigationView: NavigationView) {
    navigationView.customizeViewBinders {
        infoPanelBinder = CustomInfoPanelBinder()
    }
    navigationView.customizeViewStyles {
        infoPanelPeekHeight = defaultInfoPanelPeekHeight(navigationView.context) + 20.dp
    }
}

class CustomInfoPanelBinder : InfoPanelBinder() {
    override fun onCreateLayout(layoutInflater: LayoutInflater, root: ViewGroup): ViewGroup =
        layoutInflater.inflate(R.layout.layout_info_panel, root, false) as ViewGroup

    override fun getHeaderLayout(layout: ViewGroup): ViewGroup? =
        layout.findViewById(R.id.infoPanelHeader)

    override fun getContentLayout(layout: ViewGroup): ViewGroup? =
        layout.findViewById(R.id.infoPanelContent)
}
```

If this is still too complex, we can simplify this even further by introducing a convenience method that will accept the `layoutId` and will call necessary customization methods.

```kotlin
fun NavigationView.customizeInfoPanel(
    @LayoutRes layoutId: Int,
    @IdRes headerLayoutId: Int = R.id.mapbox_navView_infoPanelHeader,   // a default ID exposed by libnav-ui module
    @IdRes contentLayoutId: Int = R.id.mapbox_navView_infoPanelContent, // a default  ID exposed by libnav-ui module
    @Px peekHeight: Int = ViewStyleCustomization.defaultInfoPanelPeekHeight(context)
) {
    customizeViewBinders {
        infoPanelBinder = object : InfoPanelBinder() {
            override fun onCreateLayout(
                layoutInflater: LayoutInflater,
                root: ViewGroup
            ): ViewGroup {
                return layoutInflater.inflate(layoutId, root, false) as ViewGroup
            }

            override fun getHeaderLayout(layout: ViewGroup): ViewGroup? =
                layout.findViewById(headerLayoutId)

            override fun getContentLayout(layout: ViewGroup): ViewGroup? =
                layout.findViewById(contentLayoutId)
        }
    }
    customizeViewStyles {
        infoPanelPeekHeight = peekHeight
    }
}

// usage:
navigationView.customizeInfoPanel(
    layoutId = R.layout.layout_info_panel,
    headerLayoutId = R.id.infoPanelHeader,
    contentLayoutId = R.id.infoPanelContent,
    peekHeight = ViewStyleCustomization.defaultInfoPanelPeekHeight(context) + 20.dp
)
```

File: qa-test-app/src/main/res/layout/layout_drawer_menu_nav_view_custom.xml (hunk 6 above)
Comment (tomaszrybakiewicz) [reply qa-test-app/src/main/res/layout/layout_drawer_menu_nav_view_custom.xml:81]: What if I just change the message to "Custom header view for Free Drive mode".

File: libnavui-dropin/src/main/res/layout-land/mapbox_navigation_view_layout.xml [hunk 8]
```diff
@@ -129,29 +129,16 @@
 
             </androidx.constraintlayout.widget.ConstraintLayout>
 
-            <LinearLayout
+            <FrameLayout
```
Comment (kmadsen) [inline libnavui-dropin/src/main/res/layout-land/mapbox_navigation_view_layout.xml:132]: Nice 👍

File: libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/binder/infopanel/InfoPanelBinderTest.kt [hunk 9]
```diff
@@ -22,36 +21,83 @@ import org.robolectric.RobolectricTestRunner
 internal class InfoPanelBinderTest {
 
     private lateinit var ctx: Context
-    private lateinit var sut: InfoPanelBinder
+    private lateinit var sut: StubInfoPanelBinder
```
Comment (kmadsen) [inline libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/binder/infopanel/InfoPanelBinderTest.kt:24]:
```suggestion
    private lateinit var sut: InfoPanelBinder
```

**Comment by abhishek1508** [comment]
> > 👍🏻 to adding changes that allows injecting custom info panel although the plan was to introduce this through a different ticket. It's an interesting approach to allow users to:
> > add margins
> > change background drawables and colors
> > by introducing InfoPanelBinder that would enable them to change the design. However, don't you think it's a bit too much of an ask? If I just simply want to have an info panel that does not have rounded corners, with this approach as a customer I will have to write a new InfoPanelBinder.
> > My idea was to expose a few settings that could be applied to the default info panel as stated in the OP without asking users to create a new InfoPanelBinder.
> > This doesn't mean that we revert the changes introduced in the PR already. We just add a few attributes to the info panel like margins, background drawables and background colors.
> 
> @abhishek1508 I don't think this is asking too much. I think the benefits of the introduced solution here outweigh the "ask".
> 
> With the new capability, developers will not only be able to customize the info panel as described in the attached issue, but will also be able to customize how the header and content views are laid out.
> 
> Just to show how little developers need to write, here an example.
> 
> ```
> 
> <?xml version="1.0" encoding="utf-8"?>
> <androidx.constraintlayout.widget.ConstraintLayout
>     xmlns:android="http://schemas.android.com/apk/res/android"
>     xmlns:app="http://schemas.android.com/apk/res-auto"
>     xmlns:tools="http://schemas.android.com/tools"
>     android:layout_width="match_parent"
>     android:layout_height="wrap_content"
>     android:layout_margin="10dp"
>     android:background="@drawable/bg_custom_info_panel"
>     android:elevation="0dp">
> 
>     <FrameLayout
>         android:id="@+id/infoPanelHeader"
>         android:layout_width="match_parent"
>         android:layout_height="@dimen/mapbox_infoPanel_peekHeight"
>         app:layout_constraintTop_toTopOf="parent"
>         app:layout_constraintEnd_toEndOf="parent"
>         app:layout_constraintStart_toStartOf="parent"
>         tools:background="#00ccff" />
> 
>     <FrameLayout
>         android:id="@+id/infoPanelContent"
>         android:layout_width="match_parent"
>         android:layout_height="wrap_content"
>         tools:background="#ccffcc"
>         tools:minHeight="200dp"
>         app:layout_constraintTop_toBottomOf="@id/infoPanelHeader"
>         app:layout_constraintEnd_toEndOf="parent"
>         app:layout_constraintStart_toStartOf="parent" />
> 
> </androidx.constraintlayout.widget.ConstraintLayout>
> ```
> 
> And the `CustomInfoPanelBinder` that injects and describes a customized `layout_info_panel` layout:
> 
> ```kotlin
> private fun customizeInfoPanel(navigationView: NavigationView) {
>     navigationView.customizeViewBinders {
>         infoPanelBinder = CustomInfoPanelBinder()
>     }
>     navigationView.customizeViewStyles {
>         infoPanelPeekHeight = defaultInfoPanelPeekHeight(navigationView.context) + 20.dp
>     }
> }
> 
> class CustomInfoPanelBinder : InfoPanelBinder() {
>     override fun onCreateLayout(layoutInflater: LayoutInflater, root: ViewGroup): ViewGroup =
>         layoutInflater.inflate(R.layout.layout_info_panel, root, false) as ViewGroup
> 
>     override fun getHeaderLayout(layout: ViewGroup): ViewGroup? =
>         layout.findViewById(R.id.infoPanelHeader)
> 
>     override fun getContentLayout(layout: ViewGroup): ViewGroup? =
>         layout.findViewById(R.id.infoPanelContent)
> }
> ```
> 
> 
> 
> If this is still too complex, we can simplify this even further by introducing a convenience method that will accept the `layoutId` and will call necessary customization methods.
> 
> ```kotlin
> fun NavigationView.customizeInfoPanel(
>     @LayoutRes layoutId: Int,
>     @IdRes headerLayoutId: Int = R.id.mapbox_navView_infoPanelHeader,   // a default ID exposed by libnav-ui module
>     @IdRes contentLayoutId: Int = R.id.mapbox_navView_infoPanelContent, // a default  ID exposed by libnav-ui module
>     @Px peekHeight: Int = ViewStyleCustomization.defaultInfoPanelPeekHeight(context)
> ) {
>     customizeViewBinders {
>         infoPanelBinder = object : InfoPanelBinder() {
>             override fun onCreateLayout(
>                 layoutInflater: LayoutInflater,
>                 root: ViewGroup
>             ): ViewGroup {
>                 return layoutInflater.inflate(layoutId, root, false) as ViewGroup
>             }
> 
>             override fun getHeaderLayout(layout: ViewGroup): ViewGroup? =
>                 layout.findViewById(headerLayoutId)
> 
>             override fun getContentLayout(layout: ViewGroup): ViewGroup? =
>                 layout.findViewById(contentLayoutId)
>         }
>     }
>     customizeViewStyles {
>         infoPanelPeekHeight = peekHeight
>     }
> }
> 
> // usage:
> navigationView.customizeInfoPanel(
>     layoutId = R.layout.layout_info_panel,
>     headerLayoutId = R.id.infoPanelHeader,
>     contentLayoutId = R.id.infoPanelContent,
>     peekHeight = ViewStyleCustomization.defaultInfoPanelPeekHeight(context) + 20.dp
> )
> ```

I understand. But I still have the question as to why would I need to write a layout file just to be able to change the corner radius or change the background color for `InfoPanel`? `NavigationView` is offering so many styleable properties via `ViewStyleCustomization`. Why not just expose one for `InfoPanel` as well? This allows us to position ourselves that `InfoPanel` does support changing a couple of styleable properties out of the box. If you want to do more create your own `InfoPanelBinder`. 

Again, I am not suggesting we revert `InfoPanelBinder`. This is a great addition and we should definitely have it. I am suggesting that we add a few styleable properties alongside `InfoPanelBinder`, as mentioned in the OP via `ViewStyleCustomization`.

File: libnavui-dropin/src/main/res/layout/mapbox_info_panel_layout.xml (hunk 7 above)
Comment (abhishek1508) [reply libnavui-dropin/src/main/res/layout/mapbox_info_panel_layout.xml:37]: Nice 👍🏻

File: qa-test-app/src/main/res/layout/layout_drawer_menu_nav_view_custom.xml (hunk 6 above)
Comment (abhishek1508) [reply qa-test-app/src/main/res/layout/layout_drawer_menu_nav_view_custom.xml:81]: I mean you could do that, but the message is not the only thing that made me try to pull it. I wanted to, to see what the contents have. TBH, I am not seeing the value add of this switch just yet unless you already have something in mind that I am missing.

**Review by Zayankovsky — COMMENTED** [review]
I looked at this PR closer and seems like I can simply add Search views to `InfoPanelBinder` and return `null`s in `getHeaderLayout` and `getContentLayout`, because I don't need that separation.

File: libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/binder/infopanel/InfoPanelBinderTest.kt (hunk 9 above)
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/binder/infopanel/InfoPanelBinderTest.kt:24]: This is correct. We need to be able to access StubInfoPanelBinder properties.

File: qa-test-app/src/main/res/layout/layout_drawer_menu_nav_view_custom.xml (hunk 6 above)
Comment (tomaszrybakiewicz) [reply qa-test-app/src/main/res/layout/layout_drawer_menu_nav_view_custom.xml:81]: This switch allows us testing of a custom InfoPanel content binding in any NavigationState. Regardless of toggleBottomSheetFD value.

Anyways this is our development/sandbox app. Please ignore all code being written here. It's only for testing.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewStyleCustomization.kt [hunk 10]
```diff
@@ -24,7 +26,11 @@ import com.mapbox.navigation.ui.voice.view.MapboxAudioGuidanceButton
  */
 @ExperimentalPreviewMapboxNavigationAPI
 class ViewStyleCustomization {
-
+    /**
+     * Specify info panel peek height.
+     * Use [defaultInfoPanelPeekHeight] to reset to default.
+     */
+    @Px var infoPanelPeekHeight: Int? = null
```
[the remark this answers is not in the document]
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewStyleCustomization.kt:33]: > I looked at this PR closer and seems like I can simply add Search views to InfoPanelBinder and return nulls in getHeaderLayout and getContentLayout, because I don't need that separation.

Correct. Returning `null` "tells" NavigationView that supplied layout does not support header and/or content binding, and that the application will manage those views.

> In addition to peek height, can we add an option to set bottom sheet state and to observe state changes?
Maybe we should simply expose BottomSheetBehavior.

I agree. However, this is not up to me. @abhishek1508

[the remark this answers is not in the document]
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewStyleCustomization.kt:33]: Similar to `infoPanelPeekHeight` can we also add `infoPanelBackground` and `infoPanelBackgroundColor` to this file?

[the remark this answers is not in the document]
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewStyleCustomization.kt:33]: > Similar to infoPanelPeekHeight can we also add infoPanelBackground and infoPanelBackgroundColor to this file?
That's up to you. I think `infoPanelBackgroundColor` is redundant. Developers can change it via `infoPanelBackground` drawable.

[the remark this answers is not in the document]
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewStyleCustomization.kt:33]: Fine by me. Let's have `infoPanelBackground`

**Comment by tomaszrybakiewicz** [comment]
@abhishek1508 I've addressed all feedback and this is ready for another round of review.
I've added 3 customization values to `ViewStyleCustomization` that allows changes of `marginStart`, `marginEnd` and `background` when using default InfoPanel binder.
Please let me know if anything else needs to change.

**Comment by Zayankovsky** [comment]
@tomaszrybakiewicz can you add an option (maybe in a separate PR) to let users draw info panel behind the navigation bar if they use `InfoPanelBinder` or maybe in some other cases as well? For example in the screenshot below you can see that we draw Search UI behind the navigation bar, which gives a better understanding that there is content that can be scrolled up. Of course this will make users handle window insets themselves, but that is expected.

**Comment by tomaszrybakiewicz** [comment]
> @tomaszrybakiewicz can you add an option (maybe in a separate PR) to let users draw info panel behind the navigation bar if they use InfoPanelBinder or maybe in some other cases as well? For example in the screenshot below you can see that we draw Search UI behind the navigation bar, which gives a better understanding that there is content that can be scrolled up. Of course this will make users handle window insets themselves, but that is expected.

@Zayankovsky Please create a ticket and add it to the backlog. @abhishek1508 will review and prioritize this feature request accordingly.

**Comment by Zayankovsky** [comment]
> @Zayankovsky Please create a ticket and add it to the backlog. @abhishek1508 will review and prioritize this feature request accordingly.

#6073

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/infopanel/InfoPanelComponent.kt [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+import android.view.ViewGroup
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.dropin.NavigationViewContext
+import com.mapbox.navigation.ui.base.lifecycle.UIComponent
+import kotlinx.coroutines.flow.combine
+
+@ExperimentalPreviewMapboxNavigationAPI
+internal class InfoPanelComponent internal constructor(
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/infopanel/InfoPanelComponent.kt:11]:
```suggestion
internal class InfoPanelComponent constructor(
```

File: qa-test-app/src/main/res/drawable/bg_custom_info_panel2.xml [hunk 12]
```diff
@@ -0,0 +1,4 @@
+<shape xmlns:android="http://schemas.android.com/apk/res/android" android:shape="rectangle">
+  <corners android:topLeftRadius="20dp" android:topRightRadius="20dp" />
```
Comment (abhishek1508) [inline qa-test-app/src/main/res/drawable/bg_custom_info_panel2.xml:2]: Maybe we can have smaller radius values - 5dp. Right now when toggling the style switch there is literally no difference in the background except for the margins.

File: qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt [hunk 13]
```diff
@@ -236,6 +257,48 @@ class MapboxNavigationViewCustomizedActivity : DrawerActivity() {
...
+                else defaultInfoPanelPeekHeight(applicationContext)
+        }
+    }
+
+    private fun toggleCustomInfoPanelStyles(enabled: Boolean) {
+        binding.navigationView.customizeViewStyles {
+            if (enabled) {
+                infoPanelBackground = R.drawable.bg_custom_info_panel2
+                infoPanelMarginStart = 5.dp
+                infoPanelMarginEnd = 5.dp
```
Comment (abhishek1508) [inline qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt:293]:
```suggestion
                infoPanelMarginStart = 10.dp
                infoPanelMarginEnd = 10.dp
```
So that it's much more visible.

**Review by abhishek1508 — COMMENTED** [review]
Thanks for addressing the feedback @tomaszrybakiewicz. I cannot reproduce the leaks anymore. Left a couple of nit-picks

**Review by abhishek1508 — APPROVED** [review]
